azkaban-exec-server/src/test/java/azkaban/dag/DagServiceTest.java

```diff
@@ -0,0 +1,306 @@
+    final Node bNode = createNodeAndAddToTestFlow("b");
+    final Node cNode = createNodeAndAddToTestFlow("c");
+    aNode.addChildren(bNode, cNode);
+
+    addToExpectedSequence("fa", Status.RUNNING);
+    addToExpectedSequence("a", Status.RUNNING);
+    addToExpectedSequence("a", Status.SUCCESS);
+    addToExpectedSequence("b", Status.RUNNING);
+    addToExpectedSequence("c", Status.RUNNING);
+    addToExpectedSequence("b", Status.SUCCESS);
+    addToExpectedSequence("c", Status.SUCCESS);
+    addToExpectedSequence("fa", Status.SUCCESS);
+
+    runAndVerify();
+
+  }
+
+  /**
+   * Tests a DAG with one node which will fail.
+   */
+  @Test
+  public void oneNodeFailure() throws Exception {
+    final Node aNode = createNodeAndAddToTestFlow("a");
+    this.nodesToFail.add(aNode);
+    addToExpectedSequence("fa", Status.RUNNING);
+    addToExpectedSequence("a", Status.RUNNING);
+    addToExpectedSequence("a", Status.FAILURE);
+    addToExpectedSequence("fa", Status.FAILURE);
+
+    runAndVerify();
+  }
+
+  /**
+   * Tests a DAG with two nodes, fails the first one.
+   *
+   * Expects the child node to be marked canceled.
+   *
+   * a (fail)
+   * |
+   * b
+   */
+  @Test
+  public void twoNodesFailFirst() throws Exception {
+    final Node aNode = createNodeAndAddToTestFlow("a");
+    final Node bNode = createNodeAndAddToTestFlow("b");
+    aNode.addChild(bNode);
+    this.nodesToFail.add(aNode);
+
+    addToExpectedSequence("fa", Status.RUNNING);
+    addToExpectedSequence("a", Status.RUNNING);
+    addToExpectedSequence("a", Status.FAILURE);
+    addToExpectedSequence("b", Status.CANCELED);
+    addToExpectedSequence("fa", Status.FAILURE);
+
+    runAndVerify();
+  }
+
+  /**
+   * Tests a DAG with three nodes with one failure.
+   *
+   * Expects the sibling nodes to finish.
+   *
+   * <pre>
+   *       a
+   *   /      \
+   * b (fail)    c
+   * </pre>
+   */
+  @Test
+  public void threeNodesFailSecond() throws Exception {
+    final Node aNode = createNodeAndAddToTestFlow("a");
+    final Node bNode = createNodeAndAddToTestFlow("b");
+    final Node cNode = createNodeAndAddToTestFlow("c");
+    aNode.addChildren(bNode, cNode);
+
+    this.nodesToFail.add(bNode);
+
+    addToExpectedSequence("fa", Status.RUNNING);
+    addToExpectedSequence("a", Status.RUNNING);
+    addToExpectedSequence("a", Status.SUCCESS);
+    addToExpectedSequence("b", Status.RUNNING);
+    addToExpectedSequence("c", Status.RUNNING);
+    addToExpectedSequence("b", Status.FAILURE);
+    addToExpectedSequence("c", Status.SUCCESS);
+    addToExpectedSequence("fa", Status.FAILURE);
+
+    runAndVerify();
+
+  }
+
+  /**
+   * Tests a DAG with one subflow, all successful.
+   *
+   * <pre>
+   *   sfb ( subflow. "sb" stands for sub flow. This prefix is used in a node name.)
+   *   |
+   *   c
+   *
+   * subflow:
+   * a b
+   * </pre>
+   */
+  @Test
+  public void simple_subflow_success_case() throws Exception {
```

Comment:
Why no camel case? :)

Reply:
Because camel case doesn't' work well with long names. Google style guide permits it.

Reply:
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names
allows such usage:
 
“Underscores may appear in JUnit test method names to separate logical components of the name, with each component written in lowerCamelCase. One typical pattern is <methodUnderTest>_<state>, for example pop_emptyStack. There is no One Correct Way to name test methods.
“